sys/net/application_layer/dhcpv6/client.c

```diff
@@ -716,6 +848,100 @@ static bool _parse_reply(uint8_t *rep, size_t len)
+                                }
+                                break;
+                            }
+                            default:
+                                break;
+                        }
+                    }
+                    na_t1 = byteorder_ntohl(ia_na->t1);
+                    na_t2 = byteorder_ntohl(ia_na->t2);
+                    if ((na_t1 != 0) && (na_t2 != 0) &&
+                        ((server.t1 == 0) || (server.t1 >= na_t1)) &&
+                        ((t2 == 0) || (t2 >= na_t2))) {
+                        server.t1 = na_t1;
+                        t2 = na_t2;
+                        _schedule_t1_t2();
+                    }
+                    if ((iaaddr != NULL)) {
+                        uint32_t valid = byteorder_ntohl(iaaddr->valid);
+                        uint32_t pref = byteorder_ntohl(iaaddr->pref);
+                        ipv6_addr_t *addr = &iaaddr->addr;
+                        netif_t *netif = netif_get_by_id(lease->parent.ia_id.info.netif);
+                        char addr_str[IPV6_ADDR_MAX_STR_LEN];
+
+                        if (&lease->addr != NULL &&
+                            !ipv6_addr_equal(&lease->addr, addr)) {
+                            /* A different address has been leased to the client */
+                            netif_set_opt(netif, NETOPT_IPV6_ADDR_REMOVE, 64 << 8, &lease->addr, sizeof(lease->addr));
+                        }
+
+                        lease->leased = 1U;
+                        memcpy(&lease->addr, addr, sizeof(ipv6_addr_t));
+                        DEBUG("DHCPv6 client: ADD IP ADDRESS %s\n",
+                            ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)));
+
+                        if (netif_set_opt(netif, NETOPT_IPV6_ADDR, 64 << 8, addr, sizeof(*addr)) > 0) {
+                            DEBUG("IP ADDRESS successfully added!\n");
+                            /* update lifetime */
+                            if (valid < UINT32_MAX) { /* UINT32_MAX means infinite lifetime */
+                                /* the valid lifetime is given in seconds, but the NIB's timers work
+                                * in microseconds, so we have to scale down to the smallest
+                                * possible value (UINT32_MAX - 1). */
+                                valid = (valid > (UINT32_MAX / MS_PER_SEC)) ?
+                                            (UINT32_MAX - 1) : valid * MS_PER_SEC;
+                            }
+                            if (pref < UINT32_MAX) { /* UINT32_MAX means infinite lifetime */
+                                /* same treatment for pref */
+                                pref = (pref > (UINT32_MAX / MS_PER_SEC)) ?
+                                            (UINT32_MAX - 1) : pref * MS_PER_SEC;
+                            }
+                            lease->pref_until = pref;
+                            lease->valid_until = valid;
+                        }
```

Comment:
IMHO this whole block could go into its own function for better readability (and maybe even be subdivided into more sub-routines)

Reply:
I think this block should be a bit more readable by now. However, there might be some more refactoring possible as some functionality overlaps with the code for IA_PD.